### PR TITLE
Add installer binary to fetch images

### PR DESCRIPTION
**Pull request by taheris**
Resolves [PRO-3377](https://advancedtelematic.atlassian.net/browse/PRO-3377).

### CODE REVIEW

File: sota-client/docker/sota_provision.sh [hunk 1]
```diff
@@ -56,18 +61,19 @@ register_device() {
...
   pubkey=$(sed -e ':a' -e 'N' -e '$!ba' -e 's/\n/\\n/g' < "$out_pri.pub")
   ecus='{"ecu_serial":"'"$primary_serial"'","hardware_identifier":"'"$hardware_id"'","clientKey":{"keytype":"RSA","keyval":{"public":"'"$pubkey"'"}}}'
 
-  echo "Reading secondary ECUs from $cert_dir/$in_ecus"
-  while read -r secondary; do
-    [[ "$secondary" =~ [^\{.*\}$] ]] && ecus+=",$secondary"
+  while read -r serial hw_id; do
+    openssl genpkey -algorithm RSA -out "$serial.der" -outform DER -pkeyopt rsa_keygen_bits:2048
+    openssl rsa -pubout -in "$serial.der" -inform DER -out "$serial.pub"
+    pubkey=$(sed -e ':a' -e 'N' -e '$!ba' -e 's/\n/\\n/g' < "$serial.pub")
```
Comment (heartsucker) [inline sota-client/docker/sota_provision.sh]: What are you doing here? And can you make a note for when people look at this in the future.

File: sota-client/src/atomic.rs [hunk 2]
```diff
@@ -48,34 +37,64 @@ pub trait Bus: Send {
...
+/// A mapping from serials to the payloads to be delivered at each state.
+pub type Payloads = HashMap<String, HashMap<State, Payload>>;
 
-/// An `ImageWriter` should be returned at the `Fetch` state, while a
-/// `TufSigned` report should be returned at a `Commit` or `Abort` state.
-pub struct StepData {
-    pub report: Option<TufSigned>,
-    pub writer: Option<ImageWriter>,
+/// Data that may be delivered to a `Secondary` before executing state transition.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
```
Comment (heartsucker) [inline sota-client/src/atomic.rs]: Are you using `bincode` for this? Please tell me it's not JSON arrays of `u8`s.

File: sota-client/src/atomic.rs [hunk 3]
```diff
@@ -584,6 +593,7 @@ impl Bus for Multicast {
 }
 
 
+#[cfg(not(feature = "docker"))]
 #[cfg(test)]
```
Comment (heartsucker) [inline sota-client/src/atomic.rs]: Shouldn't this be `#[cfg(all(test, not(feauture = "docker")))]`?

File: sota-client/src/images.rs [hunk 4]
```diff
@@ -108,39 +112,88 @@ pub struct ImageWriter {
...
+        }
     }
 
-    /// Write a chunk of the image to the disk.
-    pub fn write_chunk(&mut self, data: &[u8], index: u64) -> Result<(), Error> {
-        trace!("writing chunk {} for {}", index, self.meta.image_name);
-        let path = format!("{}/{}", self.image_dir, self.meta.image_name);
-        let mut file = OpenOptions::new().write(true).open(path)?;
+    /// Write a single chunk directly to the output image.
+    pub fn write_direct(&mut self, data: &[u8], index: u64) -> Result<(), Error> {
```
Comment (heartsucker) [inline sota-client/src/images.rs]: I'm not sure what happens here if you get the chunks out of order. Would you get some sort of "index out of bounds" error? Or is the file allocated to the full size when it's initialized?

File: sota-client/src/atomic.rs (hunk 2 above)
Comment (taheris) [inline sota-client/src/atomic.rs]: Until there are custom bincode serialization and deserialization functions, yes. I'll add those in a separate PR.

File: sota-client/docker/sota_provision.sh (hunk 1 above)
Comment (taheris) [inline sota-client/docker/sota_provision.sh]: Comment added

File: sota-client/src/atomic.rs (hunk 3 above)
Comment (taheris) [inline sota-client/src/atomic.rs]: Sure. Fixed.

File: sota-client/src/images.rs (hunk 4 above)
Comment (taheris) [inline sota-client/src/images.rs]: Yes, `file.set_len` is called during the file creation.

**Comment by taheris** [comment]
This blocks some test driver fixes so will merge for now and make changes in additional PRs.

**Pull request merged**